src/smt/theory_str_noodler/regex.cpp

```diff
@@ -0,0 +1,559 @@
+            if(res.min_length > 0) {
+                res.universal = l_false;
+            }
+            return res;
+
+        } else if (m_util_s.re.is_of_pred(expression)) { // Handle of predicate.
+            util::throw_error("of predicate is unsupported");
+        } else if (m_util_s.re.is_opt(expression)) { // Handle optional.
+            SASSERT(expression->get_num_args() == 1);
+            const auto child{ expression->get_arg(0) };
+            SASSERT(is_app(child));
+            // min_length: 0 (epsilon)
+            RegexInfo res = get_regex_info(to_app(child), m_util_s, m);
+            res.min_length = 0;
+            return res;
+        } else if (m_util_s.re.is_range(expression)) { // Handle range.
+            SASSERT(expression->get_num_args() == 2);
+            const auto range_begin{ expression->get_arg(0) };
+            const auto range_end{ expression->get_arg(1) };
+            SASSERT(is_app(range_begin));
+            SASSERT(is_app(range_end));
+            const auto range_begin_value{ to_app(range_begin)->get_parameter(0).get_zstring()[0] };
+            const auto range_end_value{ to_app(range_end)->get_parameter(0).get_zstring()[0] };
+
+            // min_length: if there is some symbol in range --> min_length = 1; otherwise min_length = 0 (empty)
+            // empty:  if there is some symbol in range --> false; otherwise true
+            // universal: false
+            if(range_begin_value <= range_end_value) {
+                return RegexInfo{.min_length = 1, .universal = l_false, .empty = l_false};
+            } else {
+                return RegexInfo{.min_length = 0, .universal = l_false, .empty = l_true};
+            }
+        } else if (m_util_s.re.is_reverse(expression)) { // Handle reverse.
+            util::throw_error("reverse is unsupported");
+        } else if (m_util_s.re.is_union(expression)) { // Handle union (= or; A|B).
+            SASSERT(expression->get_num_args() == 2);
+            const auto left{ expression->get_arg(0) };
+            const auto right{ expression->get_arg(1) };
+            SASSERT(is_app(left));
+            SASSERT(is_app(right));
+            
+            // min_length: minimum of min_length of both guys
+            // empty: one of them is empty --> true; otherwise undef
+            // universal: min_length > 0 --> false; otherwise undef
+            RegexInfo res = get_regex_info(to_app(left), m_util_s, m);
+            RegexInfo uni = get_regex_info(to_app(right), m_util_s, m);
+            res.universal = l_undef;
+            res.min_length = std::min(uni.min_length, res.min_length);
+            if(res.empty == l_undef && uni.empty == l_false) {
+                res.empty = l_false;
+            } else if(res.empty == l_false && uni.empty == l_undef) {
+                res.empty = l_false;  
+            } else if(res.empty == l_false && uni.empty == l_false) {
+                res.empty = l_false;            
+            } else {
```

Comment:
You can also add check if both are empty, then result is also empty

Reply:
fixed

Reply:
It seems you already had that check before and I did not notice, now there are two same checks here 😆. You can actually just replace the whole big "if else if else..." with simple if-else, see https://github.com/VeriFIT/z3-noodler/pull/89/files#r1295766506

Reply:
ok